### ISSUE TITLE
Work on indexed accesses

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/boost-histogram#54]
There should be support for the indexed iterator. Something like this:

```python
for item in hist.indexed():
    item.index(0) # current index along first axis
    item.index(1) # current index along second axis
    item.bin(0) # current bin interval along first axis
    item.bin(1) # current bin interval along second axis
    item.content = 3 # Set and read bin contents  
```

This could provide the same shortcuts axes already provide, like `.bins()` and `.centers()`. So, for example, this would be the way you evaluate a function `f(x,y,...)` and fill a histogram with the values:

```python
for item in hist.indexed():
    item.content = f(*item.centers())
```

Also add a way to access the contents in the bins through numpy arrays:

```python
X, Y = hist.centers() # similar to mgrid output
```

This could remove the need to add indexed; the alternative would be, for functions that accept arrays and for histogram storages that can be converted to buffers (see #35 for an idea to allow the more common syntax for setting, and might support different storages):

```python
hist.view() = f(*hist.centers())
```

And, for functions that do not, but still for storages that can be converted to buffers:

```python
it = np.nditer([hist.view(), hist.centers()], [], [['readwrite'], ['readonly']])
with it:
    for (content, center) in it:
        content[...] = f(*center)
```

This would do two unnecessary copies due to the way `'readwrite'` on nditer was written. If you want bin edges, etc., that has to be added to the lists.

**Comment by HDembinski** [comment]
I am against duplicating interface. There should be only one way to do it. In case of reading/writing to the histogram cells, it should be `hist.view()` and `hist.view() = f(*hist.centers())`. If `f` is not vectorized, it can be vectorized with existing NumPy tools.

**Comment by henryiii** [comment]
Should we: drop indexed, make it hidden for now while we work on adding the missing methods, or leave it in the interface to mimic C++? I'm fine with any of those, though especially if we drop it, we need to provide make sure the array methods that produce bins, centers, etc. are in place.

**Comment by HDembinski** [comment]
I think it should be dropped or hidden. Maybe not throw away the code completely. I generally want the Python API to mirror the C++ API, but `indexed` is the wrong approach in Python, performance-wise, that's why I would make an exception here.

**Comment by HDembinski** [comment]
It is not clear what you mean by "array methods that produce bins, centers, etc. are in place.". We have those things on the axes and you can get the histogram counts via `.view()`.

**Comment by HDembinski** [comment]
I think there should be no `bh.indexed` and no `h._indexed`.

**Comment by HDembinski** [comment]
We could introduce a vectorized `.indexed` in a later version. Let's say you have a huge histogram, possibly using a mmap'ed storage which does not fit into memory all at once. Then having an iterator over this from `.indexed` would be useful, which gives you chunks of indices and cell values as numpy arrays.

**Comment by HDembinski** [comment]
By not providing `indexed` right now, we keep the option to implement something like this later. A chunked `indexed` would make a lot of sense even in Python.

**Comment by henryiii** [comment]
For a 2D histogram, we should be able to produce an array of [index, X, Y], so that x[0] is the array of centers for axis 1, etc. But I just realized this is trivial to add in pure python with numpy broadcasting and the axes methods. I'll drop indexed as soon as I have that in place.

Since `bh::indexed` exists in C++, we probably should not add one with an alternate meaning in Python.

**Comment by henryiii** [comment]
```python
X, Y = hist.centers() # similar to mgrid output
X, Y = hist.widths()
X, Y = hist.uppers()
X, Y = hist.lowers()
```

This will fully cover the functionality of indexed. (This might be best as a tuple of axis outputs, in the correct dimensions but broadcastable and 1D - the `ogrid` output instead of `mgrid`).

**Comment by HDembinski** [comment]
1) I am against having `.uppers()` and `.lowers()`, there should be only `.edges()`.
2) These methods should be properties of `hist.axes`, not on hist itself.

**Comment by henryiii** [comment]
This is implemented in 0.5.1; other than the discussion in #155.
